Add unit tests to teleport-generate-config AMI script

webvictim opened the pull request:
This PR:
- adds bash unit tests for the `teleport-generate-config` script used in our AMIs
  - these use [bats-core](https://github.com/bats-core/bats-core), which is now also installed in the buildbox
- fixes the mixed indentation in `teleport-generate-config` to be a uniform 4 spaces rather than the mixed 2/4 we had before
- makes various changes to `teleport-generate-config` to allow testing
  - adds a test mode flag to enable/disable calling the AWS metadata service and stubs the responses if needed
  - makes execution of `systemctl` commands optional in test mode
  - allows config directories/filenames to be overridden in test mode
  - checks whether the `teleport:adm` user:group pair exists before trying to run `chown`
  - removes some unnecessary YAML line breaks to make the outputted config file a little more readable and allow it to be parsed into blocks more easily
- uses the `route53_acm_domain_nlb_alias` as the public addr for kubernetes, SSH and tunnels if configured
  - if not, uses the old behaviour of using the proxy NLB's AWS-provided DNS name

This should allow us greater certainty that any future changes to the script will not break its functionality when deployed to AWS, without needing to go to the trouble of deploying in all the different supported modes.

Tested by building an AMI and deploying with Terraform in both LetsEncrypt and ACM modes on AWS natively, as well as ensuring that all the test cases pass.

webvictim committed:
Adds unit tests to teleport-generate-config AMI script

webvictim committed:
Disable pretty printing when running in CI

webvictim committed:
Add parallel running of unit tests

awly commented on assets/aws/files/bin/teleport-generate-config:
```diff
@@ -3,92 +3,204 @@ if [[ "${DEBUG}" == "true" ]]; then
 
+## Test functions
+# Returns 0 in test mode and 1 otherwise
+is_test() { if [[ "${TELEPORT_TEST_MODE}" == "true" ]]; then return 0; else return 1; fi }
```
```suggestion
is_test() { [[ "${TELEPORT_TEST_MODE}" == "true" ]] }
```

awly commented on assets/aws/files/tests/fixtures/common.bash:
```diff
@@ -0,0 +1,26 @@
+
+setup() {
+    load fixtures/test-setup
+    load fixtures/${TEST_SUITE?}
```
can test-specific fixtures be created inside the test itself?
it's easier to scroll up in a file to see input data, rather than looking in a separate file

webvictim replied:
I actually did this in the first iteration but it seemed less tidy. I see the benefit of doing it though, I'll give it a go.

awly commented on build.assets/Makefile:
```diff
@@ -183,6 +184,11 @@ test: buildbox
 
+.PHONY:test-sh
+test-sh: buildbox
+	docker run $(DOCKERFLAGS) $(BCCFLAGS) $(NOROOT) -t $(BUILDBOX) \
```
nit: does this need `BCCFLAGS`?

webvictim replied:
No!

awly commented on assets/aws/files/bin/teleport-generate-config:
```diff
@@ -3,92 +3,204 @@ if [[ "${DEBUG}" == "true" ]]; then
+    SSH_PORT=3023
+    if [[ "${TELEPORT_PROXY_SERVER_LB}" != "" ]]; then
+        # ACM
+        if [[ "${USE_ACM}" == "true" ]]; then
+            # if an alias for the ACM NLB is configured, use that for the proxy's ssh_public_addr
+            if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
+                SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
+            else
+                SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
+            fi
+        # non-ACM
+        else
+            if [[ "${TELEPORT_DOMAIN_NAME}" != "" ]]; then
+                SSH_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
+            else
+                SSH_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
+            fi
+        fi
+        # when using ACM, we have to set up a dedicated listener and target group as
+        # SSL termination is handled by the ACM ALB and it can't multiplex the tunnel connections
+        if [[ "${USE_ACM}" == "true" ]]; then
+            TUNNEL_PORT="${TUNNEL_LISTEN_PORT}"
+            if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
+                TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
+            else
+                TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
+            fi
+        # when ACM is not being used, Teleport handles SSL termination itself and so tunnel connections
+        # can be multiplexed over the same connection as regular web proxy traffic
+        # this is automatically enabled when Teleport's web and tunnel listeners are configured to use the same port
+        else
+            TUNNEL_PORT="${PASSED_EXTERNAL_TUNNEL_PORT}"
+            # check whether an external domain name is configured (HA mode)
+            # if not (starter mode), use an external hostname if configured, and fall back to public IP
+            if [[ "${TELEPORT_DOMAIN_NAME}" != "" ]]; then
+                TUNNEL_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
+            else
+                TUNNEL_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
+            fi
+        fi
```
nit: can these conditional blocks be merged?
they do the same checks and only differ in what vars they set

webvictim committed:
Set up fixtures inside individual tests

webvictim committed:
Address other PR feedback

webvictim committed:
Fix test function

webvictim committed:
Consolidate ACM checks

webvictim commented:
Fixed all feedback. Also replaced the use of `setup` with `setup_file` which speeds things up by only generating the config file once for each set of tests.

awly commented on assets/aws/files/bin/teleport-generate-config:
```diff
@@ -3,92 +3,203 @@ if [[ "${DEBUG}" == "true" ]]; then
+    if [[ "${TELEPORT_PROXY_SERVER_LB}" != "" ]]; then
+        # ACM
+        if [[ "${USE_ACM}" == "true" ]]; then
+            # if an alias for the ACM NLB is configured, use that for the proxy's ssh_public_addr
+            if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
+                SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
+            else
+                SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
+            fi
+
+            # when using ACM, we have to set up a dedicated listener and target group as
+            # SSL termination is handled by the ACM ALB and it can't multiplex the tunnel connections
+            TUNNEL_PORT="${TUNNEL_LISTEN_PORT}"
+            if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
+                TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
+            else
+                TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
+            fi
```
```suggestion
            TUNNEL_PORT="${TUNNEL_LISTEN_PORT}"
            SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
            TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_LB}"
            
            if [[ "${TELEPORT_PROXY_SERVER_NLB_ALIAS}" != "" ]]; then
                # if an alias for the ACM NLB is configured, use that for the proxy's ssh_public_addr
                SSH_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
                # when using ACM, we have to set up a dedicated listener and target group as
                # SSL termination is handled by the ACM ALB and it can't multiplex the tunnel connections
                TUNNEL_HOSTNAME="${TELEPORT_PROXY_SERVER_NLB_ALIAS}"
            fi
```

awly commented on assets/aws/files/bin/teleport-generate-config:
```diff
@@ -3,92 +3,203 @@ if [[ "${DEBUG}" == "true" ]]; then
+            fi
+        # non-ACM
+        else
+            # check whether an external domain name is configured (HA mode)
+            # if not (starter mode), use an external hostname if configured, and fall back to public IP
+            if [[ "${TELEPORT_DOMAIN_NAME}" != "" ]]; then
+                SSH_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
+            else
+                SSH_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
+            fi
+
+            TUNNEL_PORT="${PASSED_EXTERNAL_TUNNEL_PORT}"
+            # when ACM is not being used, Teleport handles SSL termination itself and so tunnel connections
+            # can be multiplexed over the same connection as regular web proxy traffic
+            # this is automatically enabled when Teleport's web and tunnel listeners are configured to use the same port
+            if [[ "${TELEPORT_DOMAIN_NAME}" != "" ]]; then
+                TUNNEL_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
+            else
+                TUNNEL_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
+            fi
+        fi
```
```suggestion
            TUNNEL_PORT="${PASSED_EXTERNAL_TUNNEL_PORT}"
            TUNNEL_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
            SSH_HOSTNAME="${TELEPORT_EXTERNAL_HOSTNAME:-$PUBLIC_IP}"
            if [[ "${TELEPORT_DOMAIN_NAME}" != "" ]]; then
                # check whether an external domain name is configured (HA mode)
                # if not (starter mode), use an external hostname if configured, and fall back to public IP
                SSH_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
                # when ACM is not being used, Teleport handles SSL termination itself and so tunnel connections
                # can be multiplexed over the same connection as regular web proxy traffic
                # this is automatically enabled when Teleport's web and tunnel listeners are configured to use the same port
                TUNNEL_HOSTNAME="${TELEPORT_DOMAIN_NAME}"
            fi
        fi
```

webvictim committed:
Apply suggestions from code review

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

webvictim commented:
@russjones Please bot approve.

webvictim committed:
Fix formatting

webvictim committed:
Add tests for license_file

webvictim committed:
Don't use parallel (it's somehow faster)

webvictim committed:
Merge branch 'master' into gus/add-ami-unit-tests

webvictim committed:
Merge branch 'master' into gus/add-ami-unit-tests

russjones reviewed (approved):
Bot.

webvictim committed:
Merge branch 'master' into gus/add-ami-unit-tests
